### PR TITLE
Refactor some types

### PR DESCRIPTION
**Pull request by thomastanck**

### CODE REVIEW

File: src/evaluator/datatypes.ts [hunk 1]
```diff
@@ -9,25 +8,26 @@ export enum EvalDataType {
 
 export interface Closure {
   variant: EvalDataType.Closure;
-  env: Environment | undefined;
+  env: Environment;
```
Comment (kwannoel) [inline src/evaluator/datatypes.ts:11]: Can I clarify why we've taken out undefined here?

File: src/evaluator/types.ts [hunk 2]
```diff
@@ -1,12 +1,12 @@
 import { Result } from '../utils';
-import { SExpr, SListStruct } from '../sexpr';
+import { SExprT } from '../sexpr';
```
Comment (kwannoel) [inline src/evaluator/types.ts:2]: `SExprT` means S-expr term?

File: src/evaluator/datatypes.ts (hunk 1 above)
Comment (thomastanck) [reply src/evaluator/datatypes.ts:11]: Environment is now the union of NonemptyEnvironment and undefined

File: src/evaluator/types.ts (hunk 2 above)
Comment (thomastanck) [reply src/evaluator/types.ts:2]: SExprT means SListStruct, i was going for a name like "generic S expression"

any ideas?

File: src/evaluator/__tests__/evaluator.ts [hunk 3]
```diff
@@ -9,11 +9,11 @@ import {
 } from '../evaluator';
 import { ok, getOk, getErr } from '../../utils';
 
-function expectJsonReadEvalPrint(j: JsonSExpr<never>, env: Environment | undefined) {
+function expectJsonReadEvalPrint(j: JsonSExpr, env: Environment) {
   return expect(jsonPrint(getOk(evaluate(jsonRead(j), env))));
 }
 
-function expectJsonReadEvalError(j: JsonSExpr<never>, env: Environment | undefined) {
```
Comment (kwannoel) [inline src/evaluator/__tests__/evaluator.ts:16]: Can I clarify why we use `never` here?

Comment (kwannoel) [reply src/evaluator/__tests__/evaluator.ts:16]: Realized this was a removal.... :3

**Review by kwannoel — APPROVED** [review]

**Pull request merged**